core/src/main/java/com/linecorp/armeria/server/annotation/JavaBeanRequestConverterFunction.java

```diff
@@ -0,0 +1,323 @@
+import com.linecorp.armeria.internal.AnnotatedHttpServiceParamUtil.EnumConverter;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.util.AsciiString;
+
+/**
+ * A default {@link RequestConverterFunction} which converts a request into a Java bean.
+ *
+ * <p>The following parameters are eligible for automatic injection into bean fields or properties:
+ * <ul>
+ *   <li>Path parameters to fields annotated by {@link Param}</li>
+ *   <li>HTTP parameters to fields annotated by {@link Param}</li>
+ *   <li>Headers to fields annotated by {@link Header}</li>
+ * </ul>
+ */
+public class JavaBeanRequestConverterFunction implements RequestConverterFunction {
+    private static final ConcurrentMap<Class<?>, BeanParamInfo> beanParamInfoCache = new MapMaker().weakKeys()
+                                                                                                   .makeMap();
+
+    private static final ConcurrentMap<Class<?>, Boolean> ignoredResultTypeCache = new MapMaker().weakKeys()
+                                                                                                 .makeMap();
+
+    private static final String PREFIX_SETTER_METHOD_NAME = "set";
+
+    /**
+     * register and build up parameters information for {@code expectedResultType}.
+     *
+     * @param expectedResultType a candidate data type, which will converted by JavaBeanRequestConverterFunction
+     */
+    public static synchronized void register(final Class<?> expectedResultType) {
+        if (ignoredResultTypeCache.containsKey(expectedResultType) ||
+            beanParamInfoCache.containsKey(expectedResultType)) {
+            // already processed this result type
+            return;
+        }
+
+        final BeanParamInfo beanParamInfo = buildBeanParamInfo(expectedResultType);
+        if (beanParamInfo == null) {
+            ignoredResultTypeCache.put(expectedResultType, true);
+        } else {
+            beanParamInfoCache.put(expectedResultType, beanParamInfo);
+        }
+    }
+
+    private static BeanParamInfo buildBeanParamInfo(final Class<?> expectedResultType) {
+        final List<ParamInfo> paramList = new ArrayList<>();
+
+        getAllFields(expectedResultType)
+                .forEach(
+                        field -> {
+                            final Param param = field.getAnnotation(Param.class);
+                            if (param != null) {
+                                paramList.add(new ParamFieldParamInfo(field, param));
+                            }
+
+                            final Header header = field.getAnnotation(Header.class);
+                            if (header != null) {
+                                paramList.add(new HeaderFieldParamInfo(field, header));
+                            }
+                        }
+                );
+
+        getAllMethods(expectedResultType, withPrefix(PREFIX_SETTER_METHOD_NAME), withParametersCount(1))
```

Comment:
Perhaps we don't need to filter out the methods that do not start with `set`, because we only process the methods annotated with `@Param` or `@Header`?

Reply:
fixed.